### PR TITLE
Resolve intersphinx to astropy as current build

### PR DESCRIPTION
**Pull request by nstarman**
This enables astropy to use astropy intersphinx links and correctly resolve them to the version being built.
This is implemented in SkyCoord so that links to the docs works in affiliate packages.

Part of #11555 
Request Review: @mhvk @adrn @pllim @embray 

Signed-off-by: Nathaniel Starkman (@nstarman) <nstarkman@protonmail.com>

### CODE REVIEW

**Comment by nstarman** [comment]
Success! the links in SkyCoord are still live and refer to the current build.

Screen Shot 2021-05-03 at 21 37 45

File: docs/conf.py
Comment (nstarman) [inline docs/conf.py:73]: oops. should not be included. Running a local test that this isn't necessary.

Comment (nstarman) [reply docs/conf.py]: Confirmed on local build. not related. reverting it.

File: astropy/coordinates/sky_coordinate.py
Comment (pllim) [inline astropy/coordinates/sky_coordinate.py:1038]: I haven't been following the various conversation on Sphinx xref and just by looking at this diff, I don't understand why this is needed in the core repo. Why not use the built-in Sphinx `:ref:` again? What is an example use case where this would benefit affiliated package? Thanks!

Comment (nstarman) [reply astropy/coordinates/sky_coordinate.py:1038]: > Why not use the built-in Sphinx :ref: again?

The link is to a `:doc:`, not a specific section, so `:ref:` can't be used. Perhaps the link could be changed entirely, but for page references `:doc:` is the Sphinx built-in.

> What is an example use case where this would benefit affiliated package?

In this case, for subclassing SkyCoord. Both `:doc: `and `:ref: `are not inter-sphinxed automatically, so any link in the SkyCoord subclass can break when building the docs (if warnings -> errors). This PR enables astropy to set up internal links *as if* they were intersphinx links, but have them be interpreted as internal links. Affiliate packages would see these links as intersphinx links. This is quite similar to `astropy-dev`, where Astropy sees it as internal (if it's the dev build), while affiliate packages see it as an intersphinx.

Comment (pllim) [reply astropy/coordinates/sky_coordinate.py:1038]: I cannot find any documentation of `:doc:` in the official Sphinx doc (which is kinda ironic), can you please link me to its doc? Thanks!

As for `:ref:` intersphinx, I haven't seen a problem downstream where I use `~astropy.coordinates.SkyCoord` and so on. I did see a problem when the doc here refers to simply `SkyCoord` but the fix is to mention the full namespace in core API doc.

As for referring to page vs section, I don't see the point of the former. Why not have a `:ref:` at the top of the page and then use `:ref:`?

Sorry if all these had been discussed before. If so, please point me to the old discussions. Thanks for your patience!

Comment (mhvk) [reply astropy/coordinates/sky_coordinate.py:1038]: Actually, I do remember discussion quite a while ago that docstrings should use proper http links so that people who see them in an interactive session can actually follow them....

But in that respect the scheme with `:doc:` is better than a `:ref:` - with that, the user truly has no hope to find this except by going to the docstring online. Also, the `:doc:astrop` seems a nice clear marker. Plus, this scheme has the advantage over an `http` link that we'll notice breakage sooner.

Comment (mhvk) [reply astropy/coordinates/sky_coordinate.py:1038]: One question: is the `/` in front of `/coordinates` necessary? Or does it try to find the link in the local environment of the `SkyCoord` docstring rendering otherwise?

Comment (nstarman) [reply astropy/coordinates/sky_coordinate.py:1038]: @pllim 

> I cannot find any documentation of :doc: in the official Sphinx doc (which is kinda ironic), can you please link me to its doc? Thanks!

Took me a hot sec as well. https://www.sphinx-doc.org/en/master/usage/restructuredtext/roles.html?highlight=%3Adoc%3A#cross-referencing-documents

> As for :ref: intersphinx, I haven't seen a problem downstream where I use ~astropy.coordinates.SkyCoord and so on. I did see a problem when the doc here refers to simply SkyCoord but the fix is to mention the full namespace in core API doc.

That shouldn't raise an issue. The problem, which I've encountered, is that ``:ref:`` or ``:doc:`` refer to project being built, unless the link is an intersphinx. Since there is a ``:doc:`` link in SkyCoord, when SkyCoord is subclassed the methods' docstrings can move packages due to docstring inheritance. Now the ``:ref:`` and ``:doc:`` links no longer point to Astropy and so error. Intersphinx solves this problem... but it's a bad idea to intersphinx to yourself as that doesn't get the currently built version. So this PR strips intersphinx links to Astropy when being built by Astropy, but not in affiliate packages.

> As for referring to page vs section, I don't see the point of the former. Why not have a :ref: at the top of the page and then use :ref:?

I think this is a valid point. ``:doc:`` is just a convenient ``:ref:``. I think it is actually just a ref that Sphinx handles without have a :ref: at the top.
However the above description of how a ``:doc:`` can move packages by docstring inheritance equally applies to ``:ref:``, so adding intersphinx magic is still necessary.


@mhvk 
> is the / in front of /coordinates necessary? 
I think so(?) In the docs it says "For example, if the reference :doc:`parrot` occurs in the document sketches/index, then the link refers to sketches/parrot. If the reference is :doc:`/people` or :doc:`../people`, the link refers to people." So we want to start at the base of the astropy docs.

Comment (nstarman) [reply astropy/coordinates/sky_coordinate.py:1038]: Maybe I should expand on the changelog entry to talk about resolving intersphinx and how this is useful for affiliate packages?

Comment (mhvk) [reply astropy/coordinates/sky_coordinate.py:1038]: Thanks for the explanation. It makes sense now.  The whole sphinx stuff remains a bit of black magic to me... Not sure that expanding the changelog will be needed; if anything, we'd need some docs for affiliate package maintainers, but not clear how they would find that... So, let's just stick with fixing a possible bug with subclassing `SkyCoord` here.

File: docs/changes/coordinates/11690.other.rst
Comment (mhvk) [inline docs/changes/coordinates/11690.other.rst:1]: Actually, maybe do expand that it now works if `SkyCoord` is subclassed (which I think is problematic for other reasons...)

**Review by mhvk — APPROVED** [review]
I think this looks all good, now I understand it. Perhaps still worth elaborating a bit in the changelog entry? Though not a blocker, so feel free to ignore.

File: docs/changes/coordinates/11690.other.rst
Comment (nstarman) [reply docs/changes/coordinates/11690.other.rst]: Done!

"SkyCoord links now use intersphinx links for affiliate package convenience.
Outside the astropy package, subclasses of SkyCoord, with inherited docstrings,
will now link to the astropy docs instead of raising a warning / error."

File: astropy/coordinates/sky_coordinate.py
Comment (nstarman) [reply astropy/coordinates/sky_coordinate.py:1038]: Allowing a project to refer to itself via intersphinx seems generically useful for protecting from docstring inheritance. I wish sphinx did this automatically.

Comment (nstarman) [reply astropy/coordinates/sky_coordinate.py:1038]: https://github.com/sphinx-doc/sphinx/issues/9169

**Comment by mhvk** [comment]
@pllim - what do you think, does it make sense to do this?

File: astropy/coordinates/sky_coordinate.py
Comment (pllim) [reply astropy/coordinates/sky_coordinate.py:1038]: Ah, I think I get it now. I was thinking about `:ref:` in docstring pointing to another API, but here you are talking about human doc.

Is this PR standalone or it depends on the other PR you opened over at `sphinx-astropy`?

File: docs/changes/coordinates/11690.other.rst
Comment (pllim) [inline docs/changes/coordinates/11690.other.rst:1]:
```suggestion
Sphinx RST links in ``SkyCoord`` now use intersphinx for downstream packages' convenience.
```

File: docs/changes/coordinates/11690.other.rst
Comment (pllim) [inline docs/changes/coordinates/11690.other.rst:2]:
```suggestion
Outside the ``astropy`` package, subclasses of ``SkyCoord``, with inherited docstrings,
```

File: docs/changes/coordinates/11690.other.rst
Comment (pllim) [inline docs/changes/coordinates/11690.other.rst:3]:
```suggestion
will now link to the ``astropy`` docs instead of raising a warning or an error.
```

**Review by pllim — COMMENTED** [review]
Some nitpicks on the change log but I'll leave it to your judgement which suggested changes make sense to you.

Probably also want to add a comment on the affected code about the Sphinx issue you raised. If this is implemented upstream, then we can remove it in the future.

I would still like Madison to have a look as they implemented the original function you are adding on to.

Thanks!

File: astropy/coordinates/sky_coordinate.py
Comment (nstarman) [reply astropy/coordinates/sky_coordinate.py:1038]: This PR is standalone in that it works and does the docs stuff independently.

It's definitely part of a larger push to make intersphinx and particularly xref useful for downstream packages. In some respects, this PR is the foundation on which future PRs can be built since all the downstream packages need astropy xref links to be intersphinxed, but without this PR, those links wouldn't work inside of Astropy. Now sphinx-astropy will work equally well for both Astropy and downstream packages.

**Comment by nstarman** [comment]
> Probably also want to add a comment on the affected code about the Sphinx issue you raised. If this is implemented upstream, then we can remove it in the future.

Agreed. I've added a little note on lines 375-377

> I would still like Madison to have a look as they implemented the original function you are adding on to.

👍

**Comment by embray** [comment]
How many other places in the astropy docs use `:doc:` and is there a reason we couldn't just replace them with normal `:ref:` links?

IIUC though that wouldn't solve the problem you're trying to solve anyways.

But then why single out `SkyCoord`?  In principle there are tons of other classes containing docstrings that link to other pages in the astropy docs.  And this new special feature isn't really documented anywhere, so how would remind developers writing new docstrings that it should be used?

I'm :+1: to solving this problem, and don't have any objections to this approach.  But ISTM like a deeper problem related to Sphinx...

**Comment by mhvk** [comment]
Agreed on a more general solution. Though there may not be that many other instances: I think before we really tried hard to avoid having sphinx links in docstrings, since those are not useful when doing interactive work; although hard-coded URLs are also far from ideal....

**Comment by embray** [comment]
I have in the past toyed with the idea of having a packaging hook that would go through and render all sphinx links in docstrings to URLs.  This mapping would in turn be used by an extension to IPython's help system to replace the links.

I have a little utility function called `desphinxify` which strips some other Sphinx markup from docstrings and makes them look a bit nicer, but it doesn't handle links yet.

**Comment by nstarman** [comment]
> How many other places in the astropy docs use `:doc:` and is there a reason we couldn't just replace them with normal `:ref:` links?
> IIUC though that wouldn't solve the problem you're trying to solve anyways.

Unfortunately, no that wouldn't solve this problem. Wish it did.

> But then why single out `SkyCoord`? In principle there are tons of other classes containing docstrings that link to other pages in the astropy docs.

Necessity as the mother of invention. I just first encountered this in SkyCoord. Doing a search for `:doc:`, SkyCoord is currently the only docstring that has an astropy doc link. There are a number of `:ref:` links to specific sections. I'll attend to those shortly.

> And this new special feature isn't really documented anywhere, so how would remind developers writing new docstrings that it should be used?

Hmm. good point. I'll add something to the developer docs.

> I'm 👍 to solving this problem, and don't have any objections to this approach. But ISTM like a deeper problem related to Sphinx...

Agreed that an upstream fix would be nice. I've raised an issue at https://github.com/sphinx-doc/sphinx/issues/9169. If that is implemented then the docstring changes here would be kept the same and only the intersphinx override magic in ``conf.py`` will be deprecated.

**Comment by nstarman** [comment]
Ok. @embray I think I got all `:ref:` in the docstrings and made them "intersphinx". I didn't do the ones in the docs as those cannot be inherited. Likewise, I added to the dev docs saying that all links in the docstrings should be made intersphinxed, including Astropy's. It would be nice to have some test for this, but IDK how one would test in the core package a change intended for affiliate packages.
@mhvk, this is a somewhat large diff from what you approved, so I'm re-requesting a review.

File: docs/changes/coordinates/11690.other.rst
Comment (mhvk) [inline docs/changes/coordinates/11690.other.rst:1]: Isn't it all through astropy now? Or is `SkyCoord` still the only thing that could be subclassed and thus be wrong?

File: docs/development/docguide.rst
Comment (mhvk) [inline docs/development/docguide.rst:50]: I think this is not super helpful, maybe add an example? Also, I thought this was required only for the docstrings, not for "all documentation" (which the bullet is about), so I think you need it to one of the following bullet points (maybe the `numpydoc` one? Or just its own bullet point "References in docstrings ..."

File: examples/coordinates/plot_sgr-coordinate-frame.py
Comment (mhvk) [inline examples/coordinates/plot_sgr-coordinate-frame.py:8]: Just to make sure I understand: the changes here are because it is a .py file and someone might well copy this example to their own package, modify it, but have the links still render. Correct?

**Review by mhvk — COMMENTED** [review]
@nstarman - I think the case has gotten stronger, and thanks for sticking with it! Only one real comment, about the location of the new instructions.

File: docs/changes/coordinates/11690.other.rst
Comment (nstarman) [reply docs/changes/coordinates/11690.other.rst:1]: yes, it's now all over astropy. I'm moving the note up to the top level. I've contributed a number of such fragments, all about xref, docstrings, and the documentation. Perhaps there should be a specific category / folder for such changelog fragmets, e.g. ``documentation/``.   @saimn how feasible is this with towncrier?

File: examples/coordinates/plot_sgr-coordinate-frame.py
Comment (nstarman) [reply examples/coordinates/plot_sgr-coordinate-frame.py:8]: Yes, for these cases.
I refrained from the `docs` since generally only the example codeblocks would be copied.

File: docs/development/docguide.rst
Comment (nstarman) [reply docs/development/docguide.rst]: Done! and noted 'astropy-dev' links to the dev version of astropy.

**Comment by nstarman** [comment]
> Agreed on a more general solution. Though there may not be that many other instances: I think before we really tried hard to avoid having sphinx links in docstrings, since those are not useful when doing interactive work; although hard-coded URLs are also far from ideal....

Agreed on both points.
I do think that sphinx links in docstrings can be useful for providing a note on where to find further reading and examples. Those don't detract from looking at the docstring in an interactive environment. I like @embray's idea of hot swapping for URLs. If this were done on compile so the PIP version had URLs, Jupyter Lab makes the links live!

**Review by mhvk — APPROVED** [review]
OK, looks all good now. Thanks!

Though I'm not sure what's wrong with readthedocs; did the docs pass locally and did you check the new addition in the dev docs?

Anyway, approving under the assumption that's a false positive (I had a failure recently too that I did not understand), and that the changelog fragment is located OK.

**Comment by nstarman** [comment]
Yes to the above. Let me check again the RTD is unrelated.

File: docs/changes/coordinates/11690.other.rst
Comment (saimn) [reply docs/changes/coordinates/11690.other.rst]: It's easy to do but until now the policy has been that we do not mention documentation changes. Exceptions are possible such as for big changes like you did but I'm not sure that justifies a new category. "other" seems fine to me. @pllim @bsipocz - Thoughts ?

Comment (bsipocz) [reply docs/changes/coordinates/11690.other.rst]: I don't think this would need a changelog to begin with. And as a rule of thumb, the changelog doesn't have any links. The links tend to broke on the many year timescale, and while SkyCoord is expected not to be one of those, the blanket rule of no links works better than trying to maintain a list of OK and not OK types of links.

**Comment by pllim** [comment]
RTD warnings might be related. I don't see this in `main`.

```
WARNING: :533: (WARNING/2) Inline literal start-string without end-string.
...
<unknown>:533: WARNING: undefined label: link
<unknown>:533: WARNING: undefined label: astropy:link
```

**Comment by nstarman** [comment]
Thanks @pllim. It was the example code I added in the news fragment!

**Comment by pllim** [comment]
RTD is green now but I would still like @embray to approve if they are okay with this.

File: astropy/coordinates/builtin_frames/equatorial.py
Comment (embray) [inline astropy/coordinates/builtin_frames/equatorial.py:67]: These `astropy:astropy` references are unfortunate.  I won't ask you to change them, but I wonder why those labels contained "astropy-" in the first place.  Seems redundant.  We can deal with that in a follow-up, if at all.

**Review by embray — APPROVED** [review]
Having to put `astropy:` everywhere is quite a mess, and I think people (by which I mean me) will definitely forget to do this.  But it's also exactly what I suggested you do, so I can't complain ;)

It's an unfortunate Sphinx limitation that doesn't come up very often, but for something like Astropy it is a real problem.  Thanks for opening the issue against Sphinx.

File: astropy/coordinates/builtin_frames/equatorial.py
Comment (nstarman) [reply astropy/coordinates/builtin_frames/equatorial.py:67]: Agreed. I didn't like those either. And they're now doubly redundant.

**Pull request merged**
